### PR TITLE
Add `regexp_extract_all` scalar function

### PR DESCRIPTION
**Pull request by Tishj**
This functions exists in many other systems, but surprisingly we didn't have it yet.
We do have `regexp_extract`, and this is similar in many ways, but instead of returning a single VARCHAR
The `all` version returns a list of all the matches for a given group, in a pattern on an input string.

For an example of the differences, let's look at some examples:
```sql
SELECT regexp_extract('42', '(\d?)', 1);
----
4
```
```sql
SELECT regexp_extract_all('42', '(\d?)', 1);
----
[4, 2, ]
```
Notice the empty entry we got because of the `?` as member of the group.
Now notice the difference when the content of the group isn't optional, but instead the group itself is optional:
```sql
SELECT regexp_extract_all('42', '(\d)?', 1);
----
[4, 2, NULL]
```

This difference in behavior is copied from Presto.

### CODE REVIEW

**Comment by Tishj** [comment]
This is still missing some functionality.
- Unicode characters aren't respected properly yet when dealing with empty groups
- `^` and `$` anchors aren't respected yet
ie:
```sql
select regexp_extract_all('aaa', '^a');
----
[a]
```
Does not work yet, instead we return `[a, a, a]`

File: src/function/scalar/string/regexp/regexp_extract_all.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+	if (!pattern.Match(input, *startpos, input.size(), pattern.Anchored(), groups, ngroups + 1)) {
+		return false;
+	}
+	idx_t consumed = static_cast<size_t>(groups[0].end() - (input.begin() + *startpos));
+	if (!consumed) {
+		// Empty match found, have to manually forward the input
+		// to avoid an infinite loop
+		// FIXME: support unicode characters
+		consumed++;
```
Comment (Mytherin) [inline src/function/scalar/string/regexp/regexp_extract_all.cpp:37]: You can use `LengthFun::IsCharacter` to skip forward only entire unicode codepoints

File: test/sql/function/string/regex_extract_all.test [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+query I
+WITH t(pattern) AS (
+    VALUES
+		('(a)'),
+		('(a)(a)(a)'),
+		('()'),
+		('(a)(a)'),
+		('(a)'),
+		('(a)(b)?(a)'),
+		('(a)(a)(a)(a)(a)(a)(a)')
```
Comment (Mytherin) [inline test/sql/function/string/regex_extract_all.test:660]: Perhaps add some `NULL` values to this test as well?

File: test/sql/function/string/regex_extract_all.test [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+) AS t(str)
+----
+acd	[]	[]	[]	[cd]
+bcd	[]	[]	[]	[cd]
+abcd	[]	[abcd]	[abcd]	[abcd]
+abbcd	[]	[]	[]	[cd]
+ababcd	[ababcd]	[ababcd]	[ababcd]	[ababcd]
+ab1cd	[]	[]	[]	[cd]
+
+query IIIIII
```
Comment (Mytherin) [inline test/sql/function/string/regex_extract_all.test:524]: Can `regexp_extract_all` also support the re2 modifiers (e.g. `'i'` for case-insensitivity, etc)

File: test/sql/function/string/regex_extract_all.test [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+  REGEXP_EXTRACT_ALL(str,'[\d]{1,}')  AS m3,
+  REGEXP_EXTRACT_ALL(str,'[\d]{1,2}') AS m4,
+  REGEXP_EXTRACT_ALL(str,'[\d]{0,2}') AS m5
+FROM (
+  VALUES ('03-123-4567')
+) AS t(str)
+----
+03-123-4567	[123, 456]	[]	[03, 123, 4567]	[03, 12, 3, 45, 67]	[03, , 12, 3, , 45, 67, ]
+
+#error: ab++ is not valid regex
```
Comment (Mytherin) [inline test/sql/function/string/regex_extract_all.test:537]: Are these commented out on purpose?

File: src/function/scalar/string/regexp/regexp_extract_all.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+				result_validity.SetInvalid(row);
+				continue;
+			}
+			int32_t group_index = GetGroupIndex(args, row);
+			// Get the groups
+			ExtractSingleTuple(string, lstate.constant_pattern, group_index, lstate.group_buffer, result, row);
+		}
+	} else {
+		RegexStringPieceArgs string_pieces;
+		for (idx_t row = 0; row < tuple_count; row++) {
```
Comment (Mytherin) [inline src/function/scalar/string/regexp/regexp_extract_all.cpp:166]: Can the code for the constant and non-constant pattern be unified? It looks like there is a lot of overlap.

For example, we could do something like this:
```cpp
unique_ptr<duckdb_re2::RE2> stored_re;
if (!info.constant_pattern) {
   // compile pattern string
}
auto &re = info.constant_pattern ? lstate.constant_pattern : *stored_re;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

**Comment by pdet** [comment]
cc @cpcloud

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**